src/server/frontend_xwayland/xwayland_server.cpp

```diff
@@ -139,25 +77,72 @@ void mf::XWaylandServer::execl_xwayland(XWaylandSpawner const& spawner, int wl_c
             xwayland_path.c_str(),
             dsp_str.c_str(),
             "-rootless",
-            "-wm", wm_fd_str.c_str(),
+            "-wm", x11_fd_str.c_str(),
             "-terminate",
         };
 
     for (auto const& fd : spawner.socket_fds())
     {
-        XWaylandSpawner::set_cloexec(fd, false);
+        mf::XWaylandSpawner::set_cloexec(fd, false);
         args.push_back("-listen");
         // strdup() may look like a leak, but execvp() will trash all resource management
         args.push_back(strdup(std::to_string(fd).c_str()));
     }
 
-    args.push_back(NULL);
+    args.push_back(nullptr);
 
     execvp(xwayland_path.c_str(), const_cast<char* const*>(args.data()));
 }
 
-namespace
+auto fork_xwayland_process(
+    mf::XWaylandSpawner const& spawner,
+    std::string const& xwayland_path) -> mf::XWaylandServer::XWaylandProcess
 {
+    enum { MIR, XWAYLAND, SIZE };
+    int wayland_fds[SIZE], x11_fds[SIZE];
```

Comment:
1. Let's leave uppercase for preprocessor MACROS.
2. These names are confusing:
2.1. `wayland_fds[MIR]` and `wayland_fds[XWAYLAND]` make some sense once you know how they are used as the former is passed to the Mir Wayland server, while the latter to the Xwayland client, but I don't find the names tell me what they are for.
2.2. Maybe, `wayland_pipe[server]`/`wayland_pipe[client]' would be clearer? And maybe `xcb_pipe[server]`/`xcb_pipe[client]'?

(The names in the original code not great either.)

Reply:
I think `server` and `client` are entirely unclear. If you think about it real hard, Mir is a Wayland server/X11 client and XWayland is a X11 server/Wayland client. Trying to use that relationship for naming only complicates matters. Better to name them based on the process that will end up using them.

`*_pipe` naming is good, but I'm not going to change `x11` to `xcb`, since XCB is only used on our side and isn't relevant here.

Reply:
Well, each pipe has a client and a server end, that's no more unclear than having a Mir end and an Xwayland end. Actually, I think it more clear: for example, the server end of the Wayland pipe being connected to Mir isn't intrinsic to the pipe, just how it happens to be used. (I admit the socket pairs are not intrinsically directional, but the pipes can be treated that way.)

Reply:
Changed to client/server naming hopefully in a way a little clearer to both of us